### ISSUE TITLE
Crash when editing/opening a huge tileset atlas

### ISSUE THREAD

**Issue by rockgem** [godotengine/godot#86226]
### Tested versions

4.2.1

### System information

Windows 11 Godot 4.2.1

### Issue description

there seems to be a bug where it crashes when editing and opening a tile set with a huge set of textures

### Steps to reproduce

Create a new tileset -> Drag and drop the given atlas for editting = either going to be laggy, or crash

![13_School_32x32](https://github.com/godotengine/godot/assets/63050183/c67fff6c-9924-41ef-898e-bf688fbb71a8)

### Minimal reproduction project (MRP)
none

**Comment by rockgem** [comment]
ps. i just tested it with existing and new projects, it also sometime crashes when using relatively smaller spritesheets as well.

**Comment by ATreeShine** [comment]
1. Optimize Your Textures: Ensure that your textures are not exceeding the maximum supported texture size, which is 16384×16384 pixels. For mobile GPUs, the limit is typically 4096×4096

If your textures are larger than this, consider reducing their size or splitting them into smaller textures.

2. Import Settings: Adjust the import settings for your textures in Godot. You can set a size limit on the texture import to reduce its dimensions while preserving the aspect ratio

3 TileMap Performance: If the issue is related to the TileMap's performance, consider optimizing the TileMap. For instance, you can adjust the quadrant_size property of the TileMap to improve performance

4. Collision Shapes: If your TileMap includes collision shapes, be aware that having a separate collision shape for each tile can cause performance issues

Try to simplify the collision shapes or use fewer of them.
5. Memory Management: Ensure that your project is not running out of memory due to the large textures. This can be particularly problematic when exporting to platforms with lower memory limits, such as HTML5

**Comment by rockgem** [comment]
> 1. Optimize Your Textures: Ensure that your textures are not exceeding the maximum supported texture size, which is 16384×16384 pixels. For mobile GPUs, the limit is typically 4096×4096
> 
> If your textures are larger than this, consider reducing their size or splitting them into smaller textures.
> 
> 2. Import Settings: Adjust the import settings for your textures in Godot. You can set a size limit on the texture import to reduce its dimensions while preserving the aspect ratio
> 
> 3 TileMap Performance: If the issue is related to the TileMap's performance, consider optimizing the TileMap. For instance, you can adjust the quadrant_size property of the TileMap to improve performance
> 
> 4. Collision Shapes: If your TileMap includes collision shapes, be aware that having a separate collision shape for each tile can cause performance issues
> 
> Try to simplify the collision shapes or use fewer of them. 5. Memory Management: Ensure that your project is not running out of memory due to the large textures. This can be particularly problematic when exporting to platforms with lower memory limits, such as HTML5

thanks! i will try this approach. but does this explain how the tileset editor is laggy when editing it?

**Comment by G-Milzink** [comment]
Same issue occurs while using a texture of only 384*320 pixels. Editor consistently crashes when creating new tiles, editing custom-data and/or setting up terrains, alternatiove tiles etc.

**Comment by rockgem** [comment]
similar issue #63534

**Comment by crayonape** [comment]
> Same issue occurs while using a texture of only 384*320 pixels. Editor consistently crashes when creating new tiles, editing custom-data and/or setting up terrains, alternatiove tiles etc.

Same here, it's a really small texture, but the editor keeps crashing, I can't work!

**Comment by Ekaitzsegurola** [comment]
Same problem here! I was working sice 2 weeks in 4.2 estable. Since yesterday when i was to change my tileset. Didnt happen before in 4.1 (the original project version). Tried in 4.2.1 and same problem. Does not crash when painting in map, crashes when making terrain, adding new terrain layer and picking the color, when activating the tiles, sometimes directly when clicking over the node with Tilemap to edit... 

In any case same error in infinite loop: 

ERROR: Condition "p_elem->_root != this" is true. at: remove (./core/templates/self_list.h:80)

Maybe something that was changed from 4.1? Or when upgrading the project from 4.1 to 4.2? Can i follow the code will compiling to check where is the infinite loop? (I think it may be a while loop waiting until remove that never happens but cannot guess where it comes)

My greatest tileset is 1536 x 1152, so isnt big enough

On thing i did before starting crashing was change project to Render -> Compatibility and then return back to Render -> Mobile. Can be?

**Comment by Ekaitzsegurola** [comment]
> > Also the same, getting a freeze and crash especially when setting up tiles that take up multiple squares, even after reducing tileset size. Usually after only 2~3 of them.
> 
> After investigating, the cause of the crash seems to be the editor raising this error in an infinite loop until crash:
> 
> ERROR: Condition "p_elem->_root != this" is true. at: remove (./core/templates/self_list.h:80)
> 
> Seems to be some kind of thread conflict kind of error

I am loading the code of editor trying to debug (it is my first time, i do not know how). I can bet the problem is in Clear() method that gets in while loop forever. if anyone knows where is called that method

![Captura de pantalla 2023-12-28 170635](https://github.com/godotengine/godot/assets/47370674/6d29ab59-151c-4067-9acb-01173b3c55b4)

**Comment by AThousandShips** [comment]
This was noticed previously but was said to be solved, can someone please provide a minimum reproduction project to make this easier to investigate, the original issue was not reproducible almost at all so fixing or investigating it was impossible

- A small Godot project which reproduces the issue, with no unnecessary files included. Be sure to not include the `.godot` folder in the archive (but keep `project.godot`).
- Having an MRP is very important for contributors to be able to reproduce the bug in the same way that you are experiencing it. When testing a potential fix for the issue, contributors will use the MRP to validate that the fix is working as intended.
- If the reproduction steps are not project dependent (e.g. the bug is visible in a brand new project), you can write "N/A" in the field.
- Drag and drop a ZIP archive to upload it (max 10 MB). **Do not select another field until the project is done uploading.**
- **Note for C# users:** If your issue is *not* C#-specific, please upload a minimal reproduction project written in GDScript. This will make it easier for contributors to reproduce the issue locally as not everyone has a .NET setup available.

**Comment by Ekaitzsegurola** [comment]
I have the same issue, i was playing since 1 hour with a new project tilemap trying to reproduce error without luck. I am going to try what @Nualie you just did to reproduce

I tryied in version 4.2 RC2 as the link and also crashes

**Comment by Ekaitzsegurola** [comment]
I upload the project with the bad tilemap. I removed anything else from the project. You can reproduce trying to draw the terrain like in the image below:

![Captura de pantalla 2023-12-28 190558](https://github.com/godotengine/godot/assets/47370674/1925a33f-4705-46a9-87c2-18ade7fc4043)

[GodotMyCrash.zip](https://github.com/godotengine/godot/files/13788560/GodotMyCrash.zip)

**Comment by Ekaitzsegurola** [comment]
I can confirm that if the project is changed to 4.1.2 does not crash. Same for the GodotMyCrash i uploaded with minimum to reproduce. And when returning to back to 4.2.1 crashes. Something was broken between 4.1 and 4.2

An easy way to reproduce is create a new Terrain in TerrainSets of Tilemap and moving around color selector. In most cases crashes

**Comment by AThousandShips** [comment]
Can you please try with 4.1.3 as well?

**Comment by Ekaitzsegurola** [comment]
> Can you please try with 4.1.3 as well?

In 4.1.3 works fine

I am starting to thing it could have something to do with this

https://github.com/godotengine/godot/issues/85875#issuecomment-1844918886

And maybe old Tilemaps can crash for not containing something new for that improvements

My next move is to try to create a simple project with tilemap before 4.2 dev5 and see if crashes in dev5 and after

@Nualie You said the project was from start 4.2, but maybe 4.2 dev?

**Comment by AThousandShips** [comment]
I will try to take a proper look at this in the next few days, I was investigating the original issue for some time but since it didn't have an MRP reproducing it reliably was impossible and therefore no fix was possible, this should help, I am 90% sure this is due to either some invalid copying of the `SelfList`, or some other memory corruption, either of which would break this as it would make the element no longer point to the correct list while still being seen as part of that list

I'd appreciate if someone who can reproduce this can test this PR, as it fixes one of the possible causes of this, if nothing else it narrows it down, and I will investigate it when I find the time:
* https://github.com/godotengine/godot/pull/85180

**Comment by Ekaitzsegurola** [comment]
> I will try to take a proper look at this in the next few days, I was investigating the original issue for some time but since it didn't have an MRP reproducing it reliably was impossible and therefore no fix was possible, this should help, I am 90% sure this is due to either some invalid copying of the `SelfList`, or some other memory corruption, either of which would break this as it would make the element no longer point to the correct list while still being seen as part of that list
> 
> I'd appreciate if someone who can reproduce this can test this PR, as it fixes one of the possible causes of this, if nothing else it narrows it down, and I will investigate it when I find the time:
> 
>     * [[Core] Prevent copying of `SelfList` and `SelfList::List` #85180](https://github.com/godotengine/godot/pull/85180)

I compiled your Pull Request. Is over 4.2 RC2 right?

It does not work. I am not sure but seems to crash even faster

Same error in infinite loop

ERROR: Condition "p_elem->_root != this" is true. at: remove (./core/templates/self_list.h:80)

**Comment by AThousandShips** [comment]
Thank you, then I'll get to looking at this when I can 🙂

**Comment by Ekaitzsegurola** [comment]
I have compiled two versions: One before this commit and another after. Before works fine, after versión crashes. So, should be here

Before commit (Fine): 31a7fd1951f6e9f68faa6edd2aaff87cb922a31c
After commit (Crash):  8c1e28237bfbbd5e602b164cc1868e44f7d86ec3

https://github.com/godotengine/godot/pull/81070/commits/8c1e28237bfbbd5e602b164cc1868e44f7d86ec3#diff-9bb6e79d028fa7d506aa846b3c2bbe33893107cb3d0c9e58f0a2cae1a3ca4234

If i can help in anything do not hesitate to ask


> It's on top of `master`, so not `4.2 rc2` just the latest master
> 
> Thank you for testing, then I'll get to looking at this when I can 🙂

I have compiled two versions: One before this commit and another after. Before works fine, after versión crashes. So, should be here

Before commit (Fine): 31a7fd1951f6e9f68faa6edd2aaff87cb922a31c
After commit (Crash):  8c1e28237bfbbd5e602b164cc1868e44f7d86ec3

https://github.com/godotengine/godot/pull/81070/commits/8c1e28237bfbbd5e602b164cc1868e44f7d86ec3#diff-9bb6e79d028fa7d506aa846b3c2bbe33893107cb3d0c9e58f0a2cae1a3ca4234

If i can help in anything do not hesitate to ask

**Comment by Ekaitzsegurola** [comment]
https://github.com/godotengine/godot/pull/81070/commits/8c1e28237bfbbd5e602b164cc1868e44f7d86ec3#r1310258559

In this spot of the commit, they had deliverated about the risk of infinite loop. Lines 274 and 275 of tile_map.cpp

Sad, i do not know C++ (I work with C#)

**Comment by Ekaitzsegurola** [comment]
@AThousandShips @Nualie 

In my case seems to be a wrong Pattern. When i try to draw that pattern it crashes with this message:

![imagen](https://github.com/godotengine/godot/assets/47370674/b2ac15b7-6e18-43bc-8fa4-6d395efe2ae1)


When i remove that pattern the file reduce size a lot and opening the editor again with that Pattern removed goes smoothly with no more crash for the moment. Maybe there is the problem

**Comment by baptr** [comment]
I've done a bit of debugging on this. It looks like `TilesEditorUtils` pattern refreshes can race against cleaning up some temporary TileMapLayer. Since both use TileMapLayer::set_cell with the same tilemap, I suspect they hit some kind of race on the shared `dirty_list_element` struct:

https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/scene/2d/tile_map_layer.cpp#L2124-L2125

and ends up re-adding a cell that's being removed from a dirty list, adding it twice, or some other corrupting action that gets caught on a later [clear](https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/scene/2d/tile_map_layer.cpp#L2084)

---

Across some debugging attempts, I think I've narrowed it down a bit.

Here's a trace for one of the `TileMapLaer::set_cell` calls coming from the TileEditorUtils threads:

```
Thread 61 (Thread 0x7fff1e7fc640 (LWP 1364) "godot.linuxbsd."):
#15 0x000055555ad8143f in TileMapLayer::set_cell (this=0x7ffedcac86d0, p_coords=..., p_source_id=0, p_atlas_coords=..., p_alternative_tile=0) at scene/2d/tile_map_layer.cpp:2129                                                                                                                                               #16 0x000055555ad82398 in TileMapLayer::set_pattern (this=0x7ffedcac86d0, p_position=..., p_pattern=...) at scene/2d/tile_map_layer.cpp:2277
#17 0x000055555ad5063e in TileMap::set_pattern (this=0x7ffedca509b0, p_layer=0, p_position=..., p_pattern=...) at scene/2d/tile_map.cpp:667
#18 0x0000555559e2b4cf in TilesEditorUtils::_thread (this=0x55556e48ae10) at editor/plugins/tiles/tiles_editor_plugin.cpp:97
#19 0x0000555559e2afdc in TilesEditorUtils::_thread_func (ud=0x55556e48ae10) at editor/plugins/tiles/tiles_editor_plugin.cpp:67
...
```

vs

```
Thread 1 (Thread 0x7ffff7c9d8c0 (LWP 1300) "godot.linuxbsd."):
#0  SelfList<CellData>::List::clear (this=0x7ffedcac88f0) at ./core/templates/self_list.h:108
#1  0x000055555ad81086 in TileMapLayer::internal_update (this=0x7ffedcac86d0) at scene/2d/tile_map_layer.cpp:2087
#2  0x000055555ad4cd07 in TileMap::_internal_update (this=0x7ffedca509b0) at scene/2d/tile_map.cpp:289
#3  0x000055555ad733e6 in call_with_variant_args_helper<TileMap>(TileMap*, void (TileMap::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (p_instance=0x7ffedca509b0, p_method=(void (TileMap::*)(TileMap * const)) 0x55555ad4cc78 <TileMap::_internal_update()>, p_args=0x0, r_error=...) at ./core/variant/binder_common.h:303
#4  0x000055555ad6ebf0 in call_with_variant_args<TileMap> (p_instance=0x7ffedca509b0, p_method=(void (TileMap::*)(TileMap * const)) 0x55555ad4cc78 <TileMap::_internal_update()>, p_args=0x0, p_argcount=0, r_error=...) at ./core/variant/binder_common.h:417
#5  0x000055555ad69cc0 in CallableCustomMethodPointer<TileMap>::call (this=0x7ffedc2d64c0, p_arguments=0x0, p_argcount=0, r_return_value=..., r_call_error=...) at ./core/object/callable_method_pointer.h:98
#6  0x000055555c39a0fa in Callable::callp (this=0x55556eb4ba38, p_arguments=0x0, p_argcount=0, r_return_value=..., r_call_error=...) at core/variant/callable.cpp:57
#7  0x000055555c728584 in CallQueue::_call_function (this=0x55556098f270, p_callable=..., p_args=0x55556eb4ba50, p_argcount=0, p_show_error=false) at core/object/message_queue.cpp:221
#8  0x000055555c728d61 in CallQueue::flush (this=0x55556098f270) at core/object/message_queue.cpp:326
#9  0x000055555a457b98 in SceneTree::physics_process (this=0x55556ecd1940, p_time=0.016666666666666666) at scene/main/scene_tree.cpp:471
#10 0x00005555580801f4 in Main::iteration () at main/main.cpp:3750
#11 0x0000555557fea0c0 in OS_LinuxBSD::run (this=0x7fffffffdab0) at platform/linuxbsd/os_linuxbsd.cpp:933
#12 0x0000555557fe2268 in main (argc=4, argv=0x7fffffffe0a8) at platform/linuxbsd/godot_linuxbsd.cpp:74
```

---

On another run with some [debugging prints](https://github.com/godotengine/godot/commit/cb43944f3178cc54e9f898273b2eda720de6e772) where see an overview of the same issue.

The `[]` prefix is the thread ID. We see the TilesEditorUtils thread (26) is adding cells to the MapLayer's dirty list, while the main thread (1) is trying to clear it:

```
[26] set_cell: adding 7f0840637170 to 7f08400023b0
[26] set_cell: adding 7f0840637380 to 7f08400023b0
[26] set_cell: adding **7f0840637590** to 7f08400023b0
[26] set_cell: adding 7f08406377a0 to 7f08400023b0
...
[26] set_cell: adding 7f0840695460 to 7f08400023b0
[26] set_cell: adding 7f0840695670 to 7f08400023b0
[1] Clearing dirty list 7f08400023b0
[26] set_cell: adding 7f0840695880 to 7f08400023b0
[26] set_cell: adding 7f0840695a90 to 7f08400023b0
ERROR: Trouble clearing list 7f08400023b0: element 7f0840637590 _root = 0
   at: clear (./core/templates/self_list.h:108)
[26] set_cell: adding 7f0840695ca0 to 7f08400023b0
[26] set_cell: adding 7f0840695eb0 to 7f08400023b0
[26] set_cell: adding 7f08406960c0 to 7f08400023b0
...
================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.3.dev.custom_build (c8c483cf57a768110fce57e509f9b855e69d34b7)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f095d29f520] (??:0)
[2] SelfList<CellData>::List::clear() (/home/bpu/src/godot/./core/templates/self_list.h:108)
[3] TileMapLayer::internal_update() (/home/bpu/src/godot/scene/2d/tile_map_layer.cpp:2088)
[4] TileMap::_internal_update() (/home/bpu/src/godot/scene/2d/tile_map.cpp:288 (discriminator 3))
[5] void call_with_variant_args_helper<TileMap>(TileMap*, void (TileMap::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (/home/bpu/src/godot/./core/variant/binder_common.h:308 (discriminator 4))
[6] void call_with_variant_args<TileMap>(TileMap*, void (TileMap::*)(), Variant const**, int, Callable::CallError&) (/home/bpu/src/godot/./core/variant/binder_common.h:418)
[7] CallableCustomMethodPointer<TileMap>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/bpu/src/godot/./core/object/callable_method_pointer.h:99)
[8] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/bpu/src/godot/core/variant/callable.cpp:57)
[9] CallQueue::_call_function(Callable const&, Variant const*, int, bool) (/home/bpu/src/godot/core/object/message_queue.cpp:222)
[10] CallQueue::flush() (/home/bpu/src/godot/core/object/message_queue.cpp:328)
[11] SceneTree::physics_process(double) (/home/bpu/src/godot/scene/main/scene_tree.cpp:473)
[12] Main::iteration() (/home/bpu/src/godot/main/main.cpp:3750)
[13] OS_LinuxBSD::run() (/home/bpu/src/godot/platform/linuxbsd/os_linuxbsd.cpp:933)
[14] ./bin/godot.linuxbsd.editor.dev.x86_64(main+0x19f) [0x565276102268] (/home/bpu/src/godot/platform/linuxbsd/godot_linuxbsd.cpp:76)
[15] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f095d286d90] (??:0)
[16] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f095d286e40] (??:0)
[17] ./bin/godot.linuxbsd.editor.dev.x86_64(_start+0x25) [0x565276102005] (??:?)
-- END OF BACKTRACE --
================================================================
```

**Comment by baptr** [comment]
Here's a possibly slightly too small MRP. It seems harder to trigger quickly with small/simple tilemaps, but with some patience I can reliably hit it with this:
[issue_86226.zip](https://github.com/godotengine/godot/files/13852859/issue_86226.zip) [assets src](https://anokolisa.itch.io/dungeon-crawler-pixel-art-asset-pack)

Steps:
1. Open main.tscn and select the TileMap node
2. In the TileMap tab, it helps to look switch to the Patterns subtab (note there are two previews)
3. Switch to the TileSet tab, Trees.png, Select tool, click on one of the big tree tiles
4. Expand the Rendering section and click and drag in one of the coordinate spinners (I tend to use Texture Origin y or y sort origin)
5. Wiggle your mouse for a few minutes to rapidly re-dirty that cell
   * Sometimes switching back to the TileMap tab and looking at the patterns has triggered it. Other times the previews don't redraw

Usually within ~30s, the editor should stop responding. If you have a terminal visible you should see the 
```
ERROR: Condition "p_elem->_root != this" is true.
   at: remove (./core/templates/self_list.h:80)
```

spam until it fills up the message buffer and crashes

**Comment by AThousandShips** [comment]
Thank you for those pointers, will investigate soon when I find the time

**Comment by baptr** [comment]
Probably because

https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/editor/plugins/tiles/tiles_editor_plugin.cpp#L96

triggers the async _internal_update

https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/scene/2d/tile_map_layer.cpp#L2043

via

https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/scene/2d/tile_map.cpp#L312

while https://github.com/godotengine/godot/blob/c8c483cf57a768110fce57e509f9b855e69d34b7/editor/plugins/tiles/tiles_editor_plugin.cpp#L97 is iterating over the cells

**Comment by baptr** [comment]
There may be more general purpose fixes, but https://github.com/godotengine/godot/pull/87458 made sense to me and seems to prevent this

**Comment by AThousandShips** [comment]
Investigated and this particular issue is caused by the preview generation being run on a thread and causing race conditions, there is possibly a broader thread safety concern with `TileMap` due to some race conditions when the node is not in the tree which warrants their own investigation as well

**Comment by AThousandShips** [comment]
To ensure more general thread safety I would suggest possibly not doing internal updates when not on the tree, this would make it possible to ensure thread safety, but will have to investigate further if this affects usability in general, but would probably be a separate PR unless there's no impact on the results as it'd break compat at least to some degree and this fix would be good to get for the editor crash in a future 4.2 version
